src/celeritas/neutron/interactor/NeutronElasticInteractor.hh

```diff
@@ -0,0 +1,155 @@
+#pragma once
+
+#include "corecel/Macros.hh"
+#include "corecel/Types.hh"
+#include "corecel/math/ArrayOperators.hh"
+#include "celeritas/Constants.hh"
+#include "celeritas/Quantities.hh"
+#include "celeritas/mat/IsotopeView.hh"
+#include "celeritas/neutron/data/NeutronElasticData.hh"
+#include "celeritas/phys/FourVector.hh"
+#include "celeritas/phys/Interaction.hh"
+#include "celeritas/phys/ParticleTrackView.hh"
+#include "celeritas/random/distribution/UniformRealDistribution.hh"
+
+#include "detail/InvariantQ2Sampler.hh"
+
+namespace celeritas
+{
+//---------------------------------------------------------------------------//
+/*!
+ * Perform neutron elastic scattering based on the CHIPS (Chiral invariant
+ * phase space) model.
+ *
+ * \note This performs the sampling procedure as in G4HadronElastic,
+ * G4ChipsElasticModel and G4ChipsNeutronElasticXS, as partly documented
+ * in section 21.1.3 of the Geant4 Physics Reference (release 11.2).
+ */
+class NeutronElasticInteractor
+{
+    //!@{
+    //! \name Type aliases
+    using Energy = units::MevEnergy;
+    using Mass = units::MevMass;
+    //!@}
+
+  public:
+    // Construct from shared and state data
+    inline CELER_FUNCTION
+    NeutronElasticInteractor(NeutronElasticRef const& shared,
+                             ParticleTrackView const& particle,
+                             Real3 const& inc_direction,
+                             IsotopeView const& target);
+
+    // Sample an interaction with the given RNG
+    template<class Engine>
+    inline CELER_FUNCTION Interaction operator()(Engine& rng);
+
+  private:
+    //// TYPES ////
+
+    using UniformRealDist = UniformRealDistribution<real_type>;
+
+    //// DATA ////
+
+    // Constant shared data
+    NeutronElasticRef const& shared_;
+    // Incident neutron energy
+    units::MevEnergy const inc_energy_;
+    // Incident neutron direction
+    Real3 const& inc_direction_;
+    // Target nucleus
+    IsotopeView const& target_;
+
+    // Value of neutron MeVMass
+    real_type neutron_mass_;
+    real_type neutron_p_;
+
+    // Sampler
+    UniformRealDist sample_phi_;
+    detail::InvariantQ2Sampler sample_Q2_;
+};
+
+//---------------------------------------------------------------------------//
+// INLINE DEFINITIONS
+//---------------------------------------------------------------------------//
+/*!
+ * Construct with shared and state data, and a target nucleus.
+ */
+CELER_FUNCTION NeutronElasticInteractor::NeutronElasticInteractor(
+    NeutronElasticRef const& shared,
+    ParticleTrackView const& particle,
+    Real3 const& inc_direction,
+    IsotopeView const& target)
+    : shared_(shared)
+    , inc_energy_(particle.energy())
+    , inc_direction_(inc_direction)
+    , target_(target)
+    , neutron_mass_(value_as<Mass>(shared_.neutron_mass))
+    , neutron_p_(std::sqrt(inc_energy_.value()
+                           * (inc_energy_.value() + 2 * neutron_mass_)))
```

Comment:
Since the incoming particle is a neutron, shouldn't this just be:
```suggestion
    , neutron_p_(value_as<units::MevMomentum>(particle.momentum()))
```

Reply:
Okay. I updated it accordingly and tried to use mass and momentum as much as possible instead of real_type in argument and declaration.

Reply:
It's OK to use real_type inside the class; but the class interface (constructor, call function, although not necessarily the private functions) should use units/quantities if they're not native.

Reply:
@whokion so, is `neutron_p_(value_as<units::MevMomentum>(particle.momentum()))` correct? We've already got that accessor on the `Particle` type so it's better than duplicating the calculation here.

Reply:
This was already done in the earlier commit.